fetcher: bring `curl` on par with `urllib`

alalazo opened the pull request:
This PR brings the `curl` fetch method in par with `urllib`, refactors the low level classes in `spack.fetch_strategy`, and adds type-hints.

Modifications:
* `curl` sees the same timeout, same user agent, same progress status etc. as `urllib`.
* Extracted a function to download a file to a location, which can be used outside the Spec object.
* Turned most `**kwargs` into a type list of arguments for `URLFetchStrategy`
* Removed the `spack.fetch_strategy_needs_stage` decorator
* Added type hints

**Using `curl` after the PR**

<img width="90%" height="auto" alt="Screenshot from 2025-08-29 06-45-11" src="https://github.com/user-attachments/assets/4fe3a13c-afcc-4df3-993b-d1d198ada69f" />

**Using `curl` before the PR**

<img width="80%" height="auto" alt="Screenshot from 2025-08-28 17-54-20" src="https://github.com/user-attachments/assets/59723b92-937f-4e16-9155-c0690001c994" />



<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
refactor: extract download logic in `spack.util.downloader`

Consolidates download-related functionality into a dedicated
module to improve maintainability and modularity.

Updates affected components and tests to use the new
implementation.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a timeout argument to download_file

Introduced a `timeout` parameter to the `download_file`
method.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
refactor: encapsulate URL handling in `DownloadInfo`

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
refactor: replace `extra_options` with explicit `timeout` and `cookie` fields

Encapsulate fetch options in a `TypedDict` and simplify related initialization and test logic.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
refactor: simplify fetch strategy initialization

Replaced legacy `kwargs` with explicit arguments for mirrors, expand, and extension. Removed deprecated 'checksum' from optional attributes.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Simplify `candidate_urls` implementation

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Consolidate URL fetch logic in `_fetch_from_url`

Unified fetch logic by introducing `_fetch_from_url` and
`_create_downloader` methods.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add FetchStrategy.stage property

This is the unique point to check if a stage has been set

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove `_needs_stage` decorator

Eliminated redundant `_needs_stage` decorator to simplify
fetch strategy methods, relying on `FetchStrategy.stage`
for validation instead.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Simplify `NoStageError` initialization

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Avoid dealing with circular imports here

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/downloader.py:
```diff
@@ -0,0 +1,260 @@
+            timeout: timeout in seconds for the download.
+
+        Returns:
+            DownloadInfo: An object containing details about the downloaded file,
```
```suggestion
            An object containing details about the downloaded file,
```

haampie commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -99,17 +86,26 @@ class FetchStrategy:
+            raise NoStageError()
+        return self._stage
+
+    def set_stage(self, stage) -> None:
```
why set_stage and not a setter if the thing above is a property?

haampie commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -221,112 +217,31 @@ def mirror_id(self):
-            total_bytes = None
-        return cls(total_bytes=total_bytes, enabled=enabled, get_time=get_time)
+def default_timeout() -> int:
+    if spack.config.CONFIG is None:
```
when is this ever true? it's code that exists only for tests...

alalazo committed:
refactor: replace set_stage with stage property

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Removed explicit checks for `spack.config.CONFIG` being `None`

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
> Actually, I don't think the refactor is much better than the status quo.

It depends on the definition of "much" :upside_down_face: I'll look into the additional refactor in https://github.com/spack/spack/pull/51192#issuecomment-3220244430

haampie commented:
Well, if you're looking for a common interface, I think that's it.

alalazo committed:
Treat curl the same as urllib

* Use the same progress meter
* Use the same User-Agent
* Factor common code in a unique place
* Use HTTPMessage for headers

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Make a single call to curl, instead of two

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Differentiate Windows arguments for curl

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Validate URL scheme before parsing HTTP status code

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Replace Downloader with UrlStreamReader

This gives a better API for the low-level objects reading
the file content from a URL.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Rename spack.util.downloader -> spack.util.download

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+    if url_reader is None:
+        url_reader = urllib_stream_reader()
+
+    with url_reader(url, timeout) as s:
+        partial_file = destination + ".part"
+        with open(partial_file, "wb") as f:
+            while True:
+                chunk = s.read(size=chunk_size)
+                if not chunk:
+                    break
+                f.write(chunk)
```
```python
with url_reader(url, timeout) as f, open(partial_file, "wb") as g:
    shutil.copyfileobj(f, g)
```
suffices.

Ok, maybe they don't use the "right" buffer size.

haampie replied:
Actually, I would expect the progress bar to be instantiated here, so something like:

```python
with url_reader(url, timeout) as slightly_bigger_object, open(partial_file, "wb") as g:
  data = slightly_bigger_object.f.read(...)
  progress.advance(len(data))
  ...
```

might be better

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+    def __enter__(self) -> UrlStreamReader:
+        request = urllib.request.Request(self.url, headers={"User-Agent": SPACK_USER_AGENT})
+        self._response = urlopen(request, timeout=self.timeout)
+        self._progress = FetchProgress.from_headers(
```
I will have to refactor this again for output sync. Happy to do that, or you could try to make it not instantiate this class here, but only in `download_file`.

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+        check_curl_code(self._curl_process.returncode)
+
+    def read(self, size: int) -> bytes:
+        assert self._curl_process.stdout is not None, "curl process stdout is None"
```
this assertion shouldn't be in the hot loop

alalazo replied:
Only thing we can do then is type ignore. `Popen.stdout` is `"IO[bytes] | None"`

haampie replied:
I guess in a simpler implementation the assert would be outside of the loop

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+    return DownloadInfo(request=request_info, path=destination)
+
+
+def curl_stream_reader(
```
It feels a bit over-engineered :p

alalazo committed:
Move back the progress meter in download_file

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove .cookie attribute, since too specific

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,372 @@
+
+        assert self._curl_process.stderr is not None, "curl process stderr is None"
+        while True:
+            line = self._curl_process.stderr.readline().decode("utf-8")
```
You'll have to double check, but it's likely ISO-8859-1 not UTF-8.

Why don't you use `http.client.parse_headers`? Can you check what output curl uses and if that's compatible?

Also for status follow the same logic as https://github.com/python/cpython/blob/f914e0ab09adc56abbe4bc20a195d7dc332971e1/Lib/http/client.py#L312-L323

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,372 @@
+        return self._curl_process.stdout.read(size)  # type: ignore
+
+    def request_info(self) -> RequestInfo:
+        return RequestInfo(url=self.url, effective_url=self.url, headers=self._headers)
```
leave a note that `effective_url` is not implemented?

alalazo committed:
Simplify API of download_file

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
curl: parse everything on the same stream (stdout)

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove unused function

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -168,12 +152,34 @@ def cookie_args(cookie) -> List[str]:
+            )
+
+        if 400 <= status < 600:
+            raise FetchError(
```
`DetailedHTTPError` would potentially be better/more consistent.

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -168,12 +152,34 @@ def cookie_args(cookie) -> List[str]:
-        if sys.platform == "win32":
-            return ["-D", "CON"]
-        return ["-D", "/dev/stderr"]
+    def _get_next_header(self):
```
plural? also document that curl echos the response headers of intermediate redirect responses, not just the final response.

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -168,12 +152,34 @@ def cookie_args(cookie) -> List[str]:
+            )
+
+        header = http.client.parse_headers(self._stream)
+        if self._stream.peek(len(b"HTTP/")).startswith(b"HTTP/"):
```
The actual content can start with `HTTP/`. You should branch on the current HTTP response.

alalazo replied:
> You should branch on the current HTTP response.

On which value? Are we ensured that we'll always get a `200 OK` for any successful output path?

haampie replied:
I think the correct answer is you cannot know, because what status code curl redirects for depends on the curl version.

alalazo committed:
Retrieve effective location

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
curl: treat ftp correctly

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix regression on printing attempted urls

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Branch on status code

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Raise DetailedHttpError

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+        )
+        curl_cmd = [self._curl] + curl_args
+        self._curl_process = subprocess.Popen(
+            curl_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
```
I think if we're not reading stderr it should be directed to os.devnull to avoid the deadlock of the curl process filling the stderr buffer.

alalazo replied:
Done

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+            )
+
+        if 400 <= status < 600:
+            header = http.client.parse_headers(self._stream)
```
move before the branch and remove the second call

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+            )
+
+        elif 300 <= status < 400:
+            finished = False
```
finished = 300 <= status < 400

alalazo replied:
Hmm, the fact that this does the opposite maybe suggests we should leave it 2 lines?

haampie replied:
That's a good point

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+            raise FetchError(f"Failed to fetch {self.url}: no headers returned")
+
+        self._headers = headers
+        return self
```
Did you consider return the stream object here instead of self?

alalazo replied:
Maybe the stream is not a good idea, but if `urrlib` has some guarantees on that, we can return the a `Response` object?

In that way we might get rid of some boilerplate.

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+        while True:
+            header, finished = self._get_next_header()
+            headers.append(header)
+            if finished:
```
or use `while not finished` to save a branch

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,388 @@
+        assert self._curl_process.stdout is not None, "curl process stderr is None"
+        self._stream = io.BufferedReader(self._curl_process.stdout)  # type: ignore[type-var]
+
+        # curl echoes intermediate redirect responses, so we might get multiple headers
```
you also get multiple headers without a redirect, you mean you get multiple responses, or sets of headers.

alalazo replied:
Done

alalazo committed:
Other requested changes

* stderr -> devnull
* slightly shorter while loop
* modified a word in a comment
* moved a call outside a conditional, remove redundant call

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -120,20 +120,18 @@ def __enter__(self) -> UrlStreamReader:
-        headers = []
-        while True:
+        # curl echoes intermediate redirect responses, so we might get multiple responses
+        headers, finished = [], False
```
`headers` continues to be a bad name :) anybody reading that will assume it's a list of `"Header: value"` or `("Header", "value")`.

I also don't think there's a point to store the intermediate headers, the effective URL can be obtained in the process. Plus, you wouldn't have to write `headers[-1]` and inform the reader that that indexing cannot raise.

For unit testing probably also not bad if the while loop was part of the relevant function, which can also just be a free function.

In urllib the intermediate headers are discarded as well.

alalazo replied:
Done

alalazo committed:
Other requested changes

* Use "headers" consistently for a single "response"
* Compute the effective url on the fly

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Simplify `DownloadInfo` and remove `RequestInfo` usage

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+
+
+def create_download_options(
+    method: DownloadMethod, *, extra_ars: Optional[List[str]] = None
```
```suggestion
    method: DownloadMethod, *, extra_args: Optional[List[str]] = None
```

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,381 @@
+    # Check if we somehow got an HTML file rather than the archive we
+    # asked for.  We only look at the last content type, to handle
+    # redirects properly.
+    content_types = download_info.headers.get("Content-Type")
```
this is a string, not a list of strings... so the branch below is never followed.

alalazo committed:
Refactor URL stream handling

* Replace `UrlStreamReader` classes with `UrlStream` protocol and context manager-based implementation.
* Standardize usage of `headers` and `geturl` properties.
* Fix minor typos and improve consistency in argument naming.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Restore _check_header call

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Extract a context manager to start the curl process

This can be useful for unit tests

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove boilerplate code

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a few unit tests for `curl_stream`

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+    with _start_curl_process(cmd) as curl_process:
+        scheme = urllib.parse.urlparse(url).scheme
+        assert curl_process.stdout is not None, "curl process stdout is None"
+        stream = io.BufferedReader(curl_process.stdout)  # type: ignore[type-var]
```
I think BufferedReader is not used

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+    headers = http.client.parse_headers(stream)
+    if 400 <= status < 600:
+        raise DetailedHTTPError(
+            urllib.request.Request(url), status, HTTPStatus(int(status)).phrase, headers, None
```
redundant int

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+    # Check if we somehow got an HTML file rather than the archive we
+    # asked for.  We only look at the last content type, to handle
+    # redirects properly.
+    content_types = headers.get("Content-Type")
+    if content_types and "text/html" in content_types:
```
shouldn't be plural, that's probably a remnant of the previous implementation.

```suggestion
    if "text/html" in headers.get("Content-Type", ""):
```

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+        else:
+            # curl echoes intermediate redirect responses, so we might get multiple responses
+            finished = False
+            while not finished:
+                headers, finished = _get_next_http_headers(url=url, stream=stream)
+                effective_url = headers.get("location", effective_url)
```
I would merge this into `_get_next_http_headers` so you can unit test it as a whole, without having to call `curl_stream`.

alalazo committed:
Address requests

* Remove redundant cast to int
* Remove a plural
* Remove leftover BufferedReader
* Compute the effective url in _next_headers

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -339,34 +251,52 @@ class URLFetchStrategy(FetchStrategy):
+        if fetch_options and "timeout" in fetch_options:
+            self.timeout = fetch_options["timeout"]
+
+        if (
+            fetch_options
+            and "cookie" in fetch_options
+            and self.fetch_method == DownloadMethod.CURL
+        ):
+            self.fetch_args.extend(curl_cookie_args(fetch_options["cookie"]))
```
Looks like the refactor did not accomplish its goal, cause the call site branches on the fetch method, which is what you tried to abstract away.

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+    """Returns the next headers from the stream."""
+
+    finished = True
+    status_line = stream.readline().decode("iso-8859-1")
```
This can raise two types of exceptions

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+        yield curl_process
+
+
+def _next_http_headers(url, stream) -> Tuple[http.client.HTTPMessage, str, bool]:
```
```suggestion
def _next_http_headers(url: str, stream: IO[bytes]) -> Tuple[http.client.HTTPMessage, str, bool]:
```

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+        assert curl_process.stdout is not None, "curl process stdout is None"
+        stream = curl_process.stdout
+        effective_url = url
+        if scheme not in ("http", "https"):
+            headers = HTTPMessage()
+        else:
+            # curl echoes intermediate redirect responses, so we might get multiple responses
+            finished = False
+            while not finished:
+                headers, effective_url, finished = _next_http_headers(
+                    url=effective_url, stream=stream
+                )
```
```python
if scheme in ("http", "https"):
   headers, effective_url = parse_curl_output(...)
else:
   headers = HTTPMessage()
```

haampie replied:
Actually branching on the scheme like that is wrong, because

```
$ curl -Lfi file://$PWD/hello.txt 
Content-Length: 6
Accept-ranges: bytes
Last-Modified: Tue, 02 Sep 2025 07:05:33 GMT

hello
```

Or am I missing where you pass `-i` conditionally on the scheme?

Edit: yes, I did. There should be comment here that reflects that branching on `http` and `https` is fine because we configure curl not to output headers for other schemes with `-D -`.

haampie commented on lib/spack/spack/test/download/curl.py:
```diff
@@ -0,0 +1,73 @@
+    return _mock_curl_process
+
+
+def test_curl_redirection(curl_process):
+    """Tests that the curl stream redirects to the final URL."""
+    curl_process("curl_redirection.out", return_code=0)
+
+    with spack.util.download.curl_stream(url="https://example.com/data", timeout=1) as stream:
+        assert stream.url == "https://example.com/data"
+        assert stream.geturl() == "https://archive.mesa3d.org/glu/glu-9.0.0.tar.gz"
+        assert stream.headers["Content-Type"] == "application/x-gzip"
```
This can just be a parsing test on text instead of a convoluted fake curl process

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+    destination: str,
+    timeout: int = 0,
+    chunk_size: int = 65536,
+    options: DownloadOptions = create_download_options(DownloadMethod.URLLIB),
```
this argument is hard to use from the REPL, I need to import two more symbols other than `download_file`.

I think I wouldn't mind `Literal["curl", "urllib"]`.

For cookie related stuff, did you consider `urllib.request.Request` objects instead of just `url: str`?

haampie commented on lib/spack/spack/util/web.py:
```diff
@@ -311,23 +313,24 @@ def base_curl_fetch_args(url, timeout=0):
+    scheme = urllib.parse.urlparse(url).scheme
+    curl_args = ["-f", "-L", url]  # fail on >400 errors  # resolve 3xx redirects
+
+    # Avoid getting unstructured output for FTP
```
What if curl follows a redirect to `ftp://`.

The comment should reflect the intent, namely parsing HTTP response headers. `ftp` is not special here, HTTP is.

haampie commented on lib/spack/spack/util/download.py:
```diff
@@ -0,0 +1,345 @@
+            urllib.request.Request(url), status, HTTPStatus(status).phrase, headers, None
+        )
+
+    elif 300 <= status < 400:
```
The only HTTP status code that is in the 300-399 range which is final but not a redirect is `304 Not Modified`.

It's used for conditional requests, see `spack.binary_distribution.DefaultIndexFetcher.conditional_fetch` where we do `GET ...` with `If-None-Match: "<etag>"` and you expect either 2XX with contents or `3XX` with `304 Not Modified`.

That requires action from the call site, and we don't use it with `spack.util.download` right now, but still think it's good to get it right.

Maybe look for the combination of `3XX` headers, except `304`, and ensure a `Location` header is set?
